### PR TITLE
chore: Add a read only connection for routes like Shows/NextUp

### PR DESCRIPTION
**Pull request by jaysonsantos**
This is an atempt to reduce the number of locks reported on #7237 by having a connection that can be shared among the threads without a semaphore

### CODE REVIEW

**Review by crobibero — APPROVED** [review]

**Comment by jaysonsantos** [comment]
Hey there, please, put this on hold as I think that it seems that SQLitePCL.raw is not thrad safe, I am going to give a read on its source code to see what can be done there otherwise a `System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.` happens

**Comment by nielsvanvelzen** [comment]
I marked the pull request as a draft, when it's ready again just press the "ready for review" button.

**Comment by jaysonsantos** [comment]
Ok, this should be working now.
After reading on https://github.com/ericsink/SQLitePCL.raw/issues/430 I saw that the default `SQLitePCL.pretty.ConnectionFlags` comes with `NoMutex` enabled.
In another iteration I'd add a simple "pooler" to have N read only connections and randomly distribute them to `GetConnection` calls, this way we could mitigate slow queries blocking other read queries

**Comment by jaysonsantos** [comment]
It is not perfect yet, but it seems to be a bit faster:
image
Navigation feels a bit faster as well.

**Review by 1337joe — APPROVED** [review]

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 1]
```diff
@@ -97,9 +97,16 @@ protected BaseSqliteRepository(ILogger<BaseSqliteRepository> logger)
         /// </summary>
         /// <value>The write connection.</value>
         protected SQLiteDatabaseConnection WriteConnection { get; set; }
+        protected SQLiteDatabaseConnection ReadConnection { get; set; }
```
Comment (crobibero) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs:100]:
```suggestion

        protected SQLiteDatabaseConnection ReadConnection { get; set; }
```

Newline between properties

File: Emby.Server.Implementations/Data/ManagedConnection.cs [hunk 2]
```diff
@@ -9,11 +9,12 @@ namespace Emby.Server.Implementations.Data
...
     public sealed class ManagedConnection : IDisposable
     {
-        private readonly SemaphoreSlim _writeLock;
+        private readonly SemaphoreSlim? _writeLock;
 
         private SQLiteDatabaseConnection? _db;
 
-        private bool _disposed = false;
+        private bool _disposed;
+
```
Comment (crobibero) [inline Emby.Server.Implementations/Data/ManagedConnection.cs:17]:
```suggestion
```

File: Emby.Server.Implementations/Data/ManagedConnection.cs [hunk 3]
```diff
@@ -9,11 +9,12 @@ namespace Emby.Server.Implementations.Data
...
 
         public ManagedConnection(SQLiteDatabaseConnection db, SemaphoreSlim writeLock)
```
Comment (crobibero) [inline Emby.Server.Implementations/Data/ManagedConnection.cs:19]:
```suggestion
        public ManagedConnection(SQLiteDatabaseConnection db, SemaphoreSlim? writeLock)
```

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 4]
```diff
@@ -97,9 +97,16 @@ protected BaseSqliteRepository(ILogger<BaseSqliteRepository> logger)
...
 
         protected ManagedConnection GetConnection(bool readOnly = false)
         {
+            if (readOnly)
+            {
+                ReadConnection ??= SQLite3.Open(DbFilePath, ConnectionFlags.ReadOnly, null);
+                return new ManagedConnection(ReadConnection, null!);
```
Comment (crobibero) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs:107]:
```suggestion
                return new ManagedConnection(ReadConnection, null);
```

**Review by crobibero — APPROVED** [review]

**Comment by jaysonsantos** [comment]
@crobibero how did you get to the point of that exception?
i've rebased my code with the current master and i cannot reproduce it

**Comment by crobibero** [comment]
I'm fairly certain it was during scan.
I was also informed that we had attempted this previously and ran into other issues

**Comment by jaysonsantos** [comment]
@crobibero i was taking a look to see if somehow there were more instances of the semaphore used there and when the problem happens, nothing is actually leaking, i am thinking that the problem could be with the sqlite library that is somehow not releasing the lock when disposed.
what do you say if i send a pr where I register the database connections as dependency injections and use microsoft.data.sqlite that is using newer versions of sqlitepcl.raw? the one jellyfin currentlyh uses is more than 6 years without a new release.

**Pull request merged**